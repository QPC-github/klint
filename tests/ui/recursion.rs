#![crate_type = "lib"]

use std::sync::Arc;

#[klint::preempt_count(expect = 0)]
fn might_sleep() {}

#[klint::preempt_count(expect = 0)]
fn recursive_might_sleep() {
    if false {
        recursive_might_sleep();
    }
    might_sleep();
}

fn recursive_might_sleep_unannotated() {
    if false {
        recursive_might_sleep_unannotated();
    }
    might_sleep();
}

#[klint::drop_preempt_count(expect = 0)]
struct Recursive {
    a: Option<Arc<Recursive>>,
}

impl Drop for Recursive {
    fn drop(&mut self) {
        might_sleep();
    }
}

fn drop_recur(recur: Arc<Recursive>) {}
