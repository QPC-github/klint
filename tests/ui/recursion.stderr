error: this function is recursive but preemption count expectation is not 0..
  --> $DIR/recursion.rs:16:1
   |
16 | fn recursive_might_sleep_unannotated() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expectation is inferred to be 0
   = help: try annotate the function with `#[klint::preempt_count(expect = 0)]`

error: aborting due to previous error

